lib/plan_executor/orch_client.rb

```diff
@@ -1,30 +1,36 @@
 # frozen_string_literal: true
 
+require 'jsonclient'
+
 module PlanExecutor
   class OrchClient
-    attr_reader :plan_job, :client
+    attr_reader :plan_job, :http
 
     BOLT_COMMAND_TASK = Struct.new(:name).new('bolt_shim::command').freeze
     BOLT_SCRIPT_TASK = Struct.new(:name).new('bolt_shim::script').freeze
     BOLT_UPLOAD_TASK = Struct.new(:name).new('bolt_shim::upload').freeze
 
-    def initialize(plan_job, client, logger)
+    def initialize(plan_job, http_client, logger)
       @plan_job = plan_job
-      @client = client
       @logger = logger
-      @client_url = client.config['service-url']
-      logger.debug("Creating orchestrator client for #{@client_url}")
+      @http = http_client
       @environment = 'production'
     end
 
     def finish_plan(plan_result)
-      result = @client.command.plan_finish(
+      body = {
         plan_job: @plan_job,
         result: plan_result.value || '',
         status: plan_result.status
-      )
+      }
+      response = @http.post('internal/plan_finish', body)
+      if response.status == 202
+        response.body
+      else
+        raise Bolt::Error.new(response.body['msg'], response.body['kind'], response.body['details'])
```

Comment:
Raising an error just to rescue it two lines down feels kinda gross to me?

Since a very similar pattern is used below, does it make sense to wrap this in its own function just like was done for `get`, to encapsulate this error handling? That'd clean up (or at least hide ;) ) this weirdness with an immediately-logged exception, without the need to make the two uses of `@http.post` behave totally differently

That means lines 25-31 would be replaced with a simple `post('internal/plan_finish', body)`

Reply:
I think that makes sense. I definitely don't like that it's catching its own exception, but it was either that or duplicate the `@logger.error` messages in two cases. Pulling the exception raising into a `post` method more accurately represents the desired behavior here, which is to log _any_ error, whether from an internal http client exception or from the response being an error.